### PR TITLE
NemoLite2D manual SYCL implementation

### PR DESCRIPTION
**Pull request by sergisiso**

### CODE REVIEW

**Comment by sergisiso** [comment]
@arporter This is ready for review. It is a basic implementation of SYCL that works with up-to-date versions of Intel OneAPI DPC++. It will be useful to have on master to make sure the upcoming modifications I am doing to the dl_esm_inf device pointer handling are usable by SYCL.

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/Makefile [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# Makefile for the manual SYCL version of PSyKAl'd NEMOLite2D.
+#
+# This Makefile picks up the compiler to use any flags from
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/Makefile:3]: s/use any/use plus any/

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/Makefile [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+# environment variables. e.g. to use gfortran:
+# 
+# export F90=gfortran
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/Makefile:6]: Probably these settings need to be updated (e.g. drop the OpenMP)?

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/Makefile [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+# export CXX=g++
+# export CFLAGS="-O3"
+# export F90FLAGS="-O3"
+# export OMPFLAGS=" -fopenmp"
+# export AR=ar
+
+# Additionally is uses SYCL_COMP and SYCL_FLAGS to compile the SYCL specific
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/Makefile:13]: s/is/it/

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/README.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
+# Manual SYCL version of PSyKAl NEMOLite2D #
+
+This directory contains a version of the PSyKAl form of NEMOLite2D
+that has had its PSy layer manually parallelised using SYCL.
+The Fortran(NemoLite2d) to C++/SYCL(PSy-layer) calls are done with the
+interface introduced in the `psykal_cpp` version.
+
+SYCL is a Khronos Standard for Performance Portable C++, several
+implementations of these standard are available. This code aims to be
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/README.md:9]: s/these/this/

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/README.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+will compile the default `nemolite2d_sycl.exe` binary.
+
+## Running ##
+
+Model parameters (size of domain [jpiglo,jpjglo], number of time-steps
+[nitend], whether or not and how often to do output [irecord]) may be
+configured by editing the `namelist` file.
+
+SYCL can run on several target platform (including accelerated devices),
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/README.md:39]: s/platform/platforms/

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90 [hunk 6]
```diff
@@ -0,0 +1,156 @@
+program gocean2d
+  use dl_timer, only: timer_start, timer_stop, timer_init, timer_report, i_def64
+  use grid_mod
+  use field_mod
+  use initialisation_mod, only: initialisation
+  use model_mod
+  use gocean2d_io_mod, only: model_write
+  use gocean_mod,      only: model_write_log, gocean_initialise, &
+                             gocean_finalise
+  !use likwid
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90:10]: Please rm this line.

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90 [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+  character(len=160) :: log_str
+
+  ! Initialise GOcean infrastructure
+  call gocean_initialise()
+
+  ! Create the model grid. We use a NE offset (i.e. the U, V and F
+  ! points immediately to the North and East of a T point all have the
+  ! same i,j index).  This is the same offset scheme as used by NEMO.
+  model_grid = grid_type(GO_ARAKAWA_C, &
+  !  BC_PERIODIC, BC_NON_PERIODIC ??
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90:47]: Please rm this line.

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90 [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+                         (/GO_BC_EXTERNAL,GO_BC_EXTERNAL,GO_BC_NONE/), &
+                         GO_OFFSET_NE)
+
+  !! read in model parameters and configure the model grid 
+  CALL model_init(model_grid)
+  !call likwid_markerInit()
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90:53]: Please rm.

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90 [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+  call model_write_log("((A))", TRIM(log_str))
+
+  ! Start timer for time-stepping section
+  CALL timer_start(itimer0, label='Time-stepping', &
+                   num_repeats=INT(nitend-nit000+1,kind=i_def64) )
+
+  !! time stepping 
+  do istp = nit000, nitend, 1
+
+     !call model_write_log("('istp == ',I6)",istp)
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90:101]: Please rm.

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90 [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+  ! Compute and output some checksums for error checking
+  call model_write_log("('ua checksum = ', E16.8)", &
+                       field_checksum(ua_fld))
+  call model_write_log("('va checksum = ', E16.8)", &
+                       field_checksum(va_fld))
+
+  !! finalise the model run
+  call model_finalise()
+  !call likwid_markerClose()
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90:125]: Please rm.

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/time_step_sycl.cpp [hunk 11]
```diff
@@ -0,0 +1,686 @@
...
+        for(auto dev : devices){
+            auto platform_name = dev.get_platform().get_info<info::platform::name>();
+            auto dev_name = dev.get_info<info::device::name>();
+            std::cout <<  " " << index++ << " ) " << platform_name << " : " <<
+                dev_name << std::endl;
+        }
+
+        int selected_device = 0;
+        if(const char* env_p = std::getenv("SYCL_PLATFORM")){
+            selected_device = std::stoi(env_p);
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/time_step_sycl.cpp:108]: Please could you add a check here that the selected device is valid.

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/time_step_sycl.cpp [hunk 12]
```diff
@@ -0,0 +1,686 @@
...
+#ifdef USE_TIMER
+    TimerStop();
+    TimerReport();
+#endif
+
+}
+
+extern "C" void sycl_read_from_device(void* from, double* to,
+                                        int nx, int ny, int width){
+    // Just add a wait for now (it may not work in accelerators yet)
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/time_step_sycl.cpp:684]: Please could you extend this comment a bit to explain whether or not you expect to have to add an explicit copy here in future.

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/timing.cpp [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+  }
+
+  /* Pack the times into an array of structs */
+  for(i=0; i<num_active_timers; i++){
+    timings[i].val  = timer_list[i].sum_time;
+    raw_times[i]    = timer_list[i].sum_time;
+    timings[i].rank = idPE;
+  }
+
+  /** \todo Ideally this MPI code should be in Parallel_MPI.c but
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/timing.cpp:294]: Remove this comment?

**Review by arporter — CHANGES_REQUESTED** [review]
Good job Sergi. My C++ is so out of date that I have to admit to only doing a cursory review of the C++ bits.
However, I have managed to compile and run on Glados.
There are just a few minor things to tidy up.

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/Makefile (hunk 1 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/Makefile:3]: Done

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/Makefile (hunk 2 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/Makefile:6]: Removed OpenMP and AR

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/Makefile (hunk 3 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/Makefile:13]: Fixed

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/README.md (hunk 4 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/README.md:9]: Fixed

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/README.md (hunk 5 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/README.md:39]: Fixed

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90 (hunk 6 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90:10]: Done

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90 (hunk 7 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90:47]: Done

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90 (hunk 8 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90:53]: Done

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90 (hunk 9 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90:101]: Done

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90 (hunk 10 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/nemolite2d.f90:125]: Done

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/time_step_sycl.cpp (hunk 11 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/time_step_sycl.cpp:108]: I added some basic checks and exception catching around this line

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/time_step_sycl.cpp (hunk 12 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/time_step_sycl.cpp:684]: I extended the comment.

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/timing.cpp (hunk 13 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/timing.cpp:294]: I removed it.

**Comment by sergisiso** [comment]
@arporter I made the requested modifications. Ready for next review.

**Review by arporter — APPROVED** [review]
All requested changes haven been made.
Code still compiles and runs (on Glados). New error checking works.

**Pull request merged**